source/Calamari.Shared/Integration/Packages/PackageName.cs

```diff
@@ -80,13 +78,34 @@ static bool TryParseUnsafeFileName(string fileName, out string packageId, out IV
             const string semanticVersionPattern = @"(?<semanticVersion>(\d+(\.\d+){0,3}" // Major Minor Patch
                                                   + @"(-[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*)?)" // Pre-release identifiers
                                                   + @"(\+[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*)?)"; // Build Metadata
-            const string extensionPattern = @"(?<extension>(\.([a-zA-Z0-9])+)+)"; //Extension (wont catch two part extensions like .tar.gz if there is a pre-release tag)
 
-            var match = Regex.Match(fileName, $@"^{packageIdPattern}\.{semanticVersionPattern}{extensionPattern}$", RegexOptions.IgnoreCase);
+            var fileNamePattern = $@"{packageIdPattern}\.{semanticVersionPattern}";
+
+            // here we're getting opinionated about how we split up packages that have both
+            // pre-release identifiers and/or build metadata and also a two-part extension name.
+            // for example "foo.1.0.0-release.tar.gz" could mean:
+            // - release tag: "release.tar" and extension: "gz"
+            // - release tag: "release" and extension: "tar.gz"
+            // The thinking here is
+            // - "tar.gz" and friends are extensions we know about
+            // - this is a method specific to packages and archives
+            // - cases like the above "tar" are unlikely to be part of a release tag or build metadata
+            // so the below method strips out the extensions we know about before performing
+            // the match on the filename. If we don't find any extensions we know about
+            // we fall back to trying to pattern match one in the filename.
+            if (TryMatchKnownExtensions(fileName, out var strippedFileName, out var extensionMatch))
+            {
+                fileName = strippedFileName;
+            } else { 
+                const string extensionPattern = @"(?<extension>(\.([a-zA-Z0-9])+)+)"; //Extension (wont catch two part extensions like .tar.gz if there is a pre-release tag)
+                fileNamePattern += extensionPattern;
+            } 
```

Comment:
What if we always stripped the extension off (not just `.tar.*`), then the regex to get the package ID can just be `$@"{packageIdPattern}\.{semanticVersionPattern}"`. It should make the code easier to follow

Reply:
it does, but we would miss out on any unknown two-part extensions for commonly formatted packages, i.e:

As Is

package name | .zip  .tar  .nupkg | .tar.gz  .tar.bz2 .tar.xyz | .abc.xyx 
--- | --- | --- | ---
foo.1.0.0  | ✔️ | ✔️  | ✔️ 
foo.1.0.0-tag   | ✔️ | ✔️ | ❌ 
foo.1.0.0-tag-release.tag   | ✔️ | ✔️ | ❌ 
foo.1.0.0+buildmeta   | ✔️ | ✔️ | ❌
foo.1.0.0-tag-release.tag+buildmeta   | ✔️ | ✔️ | ❌ 

If we change to make the code easier to follow

package name | .zip  .tar  .nupkg | .tar.gz  .tar.bz2 .tar.xyz | .abc.xyx 
--- | --- | --- | ---
foo.1.0.0  | ✔️ | ✔️  | ❌ 
foo.1.0.0-tag   | ✔️ | ✔️ | ❌ 
foo.1.0.0-tag-release.tag   | ✔️ | ✔️ | ❌ 
foo.1.0.0+buildmeta   | ✔️ | ✔️ | ❌
foo.1.0.0-tag-release.tag+buildmeta   | ✔️ | ✔️ | ❌

Reply:
I'm not sure how likely we are to come across new 2-part extensions, but fair point 👍